File: examples/job-interview.ts
#!/usr/bin/env ts-node
/**
 * Job Interview Example
 * =====================
 * 
 * This example demonstrates:
 * - Possible traits that activate based on conversation
 * - Confidential fields (has_mentored, shows_leadership)
 * - Conclusion fields (preparedness and cultural_fit assessment)
 * - Regular fields with validation (@must)
 * 
 * Run with:
 *   npx ts-node examples/job-interview.ts
 * 
 * For automated demo:
 *   npx ts-node examples/job-interview.ts --auto
 */

import { chatfield } from '../src/builders/gatherer-builder'
import { Interviewer } from '../src/core/interviewer'
import * as readline from 'readline'
import * as process from 'process'

/**
 * Create a job interview for a software engineer position
 */
function createJobInterview() {
  return chatfield()
    .type("JobInterview")
    .desc("Software Engineer position interview")
    
    .alice()
      .type("Hiring Manager")
      .trait.apply("Professional and encouraging")
    
    .bob()
      .type("Candidate")
      .trait.possible("career-changer", "mentions different industry or transferable skills")
      .trait.possible("senior-level", "10+ years of experience or leadership roles")
    
    .field("experience")
      .desc("Tell me about your relevant experience")
      .must("specific examples or projects")
    
    .field("technical_skills")
      .desc("What programming languages and technologies are you proficient in?")
      .hint("Please mention specific languages, frameworks, and tools")
    
    // Confidential field - tracked but never mentioned
    .field("has_mentored")
      .desc("Gives specific evidence of professionally mentoring junior colleagues")
      .confidential()
      .asBool()
    
    // Another confidential field
    .field("shows_leadership")
      .desc("Demonstrates leadership qualities or initiatives")
      .confidential()
      .asBool()
    
    // Conclusion field - assessed at the end
    .field("preparedness")
      .desc("Evaluate candidate's preparation: research, questions, examples")
      .conclude()
      .as_one.assessment("unprepared", "somewhat prepared", "well prepared", "exceptionally prepared")
    
    .field("cultural_fit")
      .desc("Assessment of cultural fit based on values and communication style")
      .conclude()
      .as_one.assessment("poor fit", "potential fit", "good fit", "excellent fit")
    
    .build()
}

/**
 * Run the interview interactively
 */
async function runInteractive(interview: any): Promise<boolean> {
  const threadId = `job-interview-${process.pid}`
  console.log(`Starting job interview (thread: ${threadId})`)
  console.log("=".repeat(60))
  console.log("Software Engineer Position Interview")
  console.log("=".repeat(60))
  
  const interviewer = new Interviewer(interview, { threadId })
  
  // Create readline interface for user input
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  })
  
  const askQuestion = (prompt: string): Promise<string> => {
    return new Promise((resolve) => {
      rl.question(prompt, (answer) => {
        resolve(answer.trim())
      })
    })
  }
  
  let userInput: string | null = null
  console.log("\n[Interview begins]\n")
  
  while (!interview._done) {
    const message = await interviewer.go(userInput)
    
    if (message) {
      console.log(`\nInterviewer: ${message}`)
    }
    
    if (!interview._done) {
      try {
        userInput = await askQuestion("\nCandidate: ")
      } catch (error) {
        console.log("\n[Interview ended by user]")
        break
      }
    }
  }
  
  if (interview._done) {
    console.log("\n[Interview complete - Thank you for your time]")
  }
  
  rl.close()
  return interview._done
}

/**
 * Run with prefab inputs for demonstration
 */
async function runAutomated(interview: any): Promise<boolean> {
  const prefabInputs = [
    "I've spent 5 years in finance, but I've been passionate about programming since college. " +
    "I built an automated trading system in Python that saved our team 20 hours per week. " +
    "I also created a risk analysis dashboard using React and D3.js.",
    
    "I'm proficient in Python, JavaScript, and SQL. I've worked extensively with React, Node.js, " +
    "and Django. I also have experience with Docker, AWS, and CI/CD pipelines using GitHub Actions.",
    
    "In my previous role, I regularly mentored junior analysts on Python programming and data analysis. " +
    "I also led a cross-functional team to implement our new reporting system.",
    
    "I've researched your company's focus on fintech innovation and your recent Series B funding. " +
    "I'm particularly excited about your API-first approach and how my background in both finance " +
    "and engineering could contribute to bridging technical and business requirements. " +
    "What are the biggest technical challenges your team is currently facing?"
  ]
  
  const threadId = `job-demo-${process.pid}`
  console.log(`Running automated demo (thread: ${threadId})`)
  console.log("=".repeat(60))
  console.log("Software Engineer Position Interview (Demo)")
  console.log("=".repeat(60))
  
  const interviewer = new Interviewer(interview, { threadId })
  
  let inputIndex = 0
  let userInput: string | null = null
  
  console.log("\n[Interview begins]\n")
  
  while (!interview._done) {
    const message = await interviewer.go(userInput)
    
    if (message) {
      console.log(`\nInterviewer: ${message}`)
    }
    
    if (!interview._done) {
      if (inputIndex < prefabInputs.length) {
        userInput = prefabInputs[inputIndex++]
        console.log(`\nCandidate: ${userInput}`)
        // Add a small delay for readability
        await new Promise(resolve => setTimeout(resolve, 1500))
      } else {
        console.log("\n[Demo inputs exhausted]")
        break
      }
    }
  }
  
  return interview._done
}

/**
 * Display interview results and assessment
 */
function displayResults(interview: any): void {
  console.log("\n" + "=".repeat(60))
  console.log("INTERVIEW ASSESSMENT")
  console.log("-".repeat(60))
  
  // Regular fields
  if (interview.experience) {
    console.log("\nExperience Summary:")
    const preview = interview.experience.length > 100 
      ? interview.experience.substring(0, 100) + "..."
      : interview.experience
    console.log(`  ${preview}`)
  }
  
  if (interview.technical_skills) {
    console.log("\nTechnical Skills:")
    const preview = interview.technical_skills.length > 100
      ? interview.technical_skills.substring(0, 100) + "..."
      : interview.technical_skills
    console.log(`  ${preview}`)
  }
  
  // Check activated traits
  const bobRole = interview._chatfield.roles.get('bob')
  const traits: string[] = []
  
  if (bobRole?.possibleTraits.get('career-changer')?.active) {
    traits.push("Career Changer")
  }
  if (bobRole?.possibleTraits.get('senior-level')?.active) {
    traits.push("Senior Level")
  }
  
  if (traits.length > 0) {
    console.log("\nCandidate Profile:")
    traits.forEach(trait => console.log(`  • ${trait}`))
  }
  
  // Confidential assessments (internal only)
  console.log("\n[Internal Assessment]")
  
  if (interview.has_mentored !== null && interview.has_mentored !== undefined) {
    const mentored = interview.has_mentored.as_bool ? "Yes" : "No"
    console.log(`  Mentoring Experience: ${mentored}`)
  }
  
  if (interview.shows_leadership !== null && interview.shows_leadership !== undefined) {
    const leadership = interview.shows_leadership.as_bool ? "Yes" : "No"
    console.log(`  Leadership Qualities: ${leadership}`)
  }
  
  // Conclusion fields
  if (interview.preparedness) {
    console.log(`  Preparedness: ${interview.preparedness.as_one_assessment || interview.preparedness}`)
  }
  
  if (interview.cultural_fit) {
    console.log(`  Cultural Fit: ${interview.cultural_fit.as_one_assessment || interview.cultural_fit}`)
  }
  
  // Calculate overall recommendation
  console.log("\n[Recommendation]")
  const hasMentored = interview.has_mentored?.as_bool || false
  const showsLeadership = interview.shows_leadership?.as_bool || false
  const preparedness = interview.preparedness?.as_one_assessment
  const culturalFit = interview.cultural_fit?.as_one_assessment
  
  let recommendation = "Needs Review"
  if (preparedness === "exceptionally prepared" && culturalFit === "excellent fit") {
    recommendation = "Strong Hire"
  } else if (
    (preparedness === "well prepared" || preparedness === "exceptionally prepared") &&
    (culturalFit === "good fit" || culturalFit === "excellent fit")
  ) {
    recommendation = hasMentored || showsLeadership ? "Hire" : "Likely Hire"
  } else if (culturalFit === "poor fit") {
    recommendation = "No Hire"
  }
  
  console.log(`  Decision: ${recommendation}`)
  console.log("=".repeat(60))
}

/**
 * Main entry point
 */
async function main() {
  // Load environment variables if .env exists
  try {
    require('dotenv').config()
  } catch {
    // .env file is optional
  }
  
  // Check for OpenAI API key
  if (!process.env.OPENAI_API_KEY) {
    console.error("Error: OPENAI_API_KEY environment variable is required")
    console.error("Please set it in your environment or create a .env file")
    process.exit(1)
  }
  
  // Parse command line arguments
  const args = process.argv.slice(2)
  const isAuto = args.includes('--auto')
  
  // Create the interview
  const interview = createJobInterview()
  
  try {
    // Run the interview
    const completed = isAuto 
      ? await runAutomated(interview)
      : await runInteractive(interview)
    
    // Display results if completed
    if (completed) {
      displayResults(interview)
    } else {
      console.log("\nInterview incomplete.")
    }
  } catch (error) {
    console.error("\nError during interview:", error)
    process.exit(1)
  }
}

// Run if executed directly
if (require.main === module) {
  main().catch(console.error)
}

export { createJobInterview, displayResults }

File: chatfield/examples/job_interview.py
#!/usr/bin/env python3
"""
Job Interview Example
=====================

This example demonstrates:
- Possible traits that activate based on conversation
- Confidential fields (has_mentored)
- Conclusion fields (preparedness assessment)
- Regular fields with validation (@must)

Run with:
    python examples/job_interview.py

For automated demo:
    python examples/job_interview.py --auto
"""

import os
import sys
import argparse
from pathlib import Path
import dotenv

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

from chatfield import chatfield
from chatfield import Interviewer


def create_job_interview():
    """Create a job interview for a software engineer position."""
    return (chatfield()
        .type("JobInterview")
        .desc("Software Engineer position interview")
        
        .alice()
            .type("Hiring Manager")
            .trait("Professional and encouraging")
        
        .bob()
            .type("Candidate")
            .trait.possible("career-changer", "mentions different industry or transferable skills")
            .trait.possible("senior-level", "10+ years of experience or leadership roles")
        
        .field("experience")
            .desc("Tell me about your relevant experience")
            .must("specific examples or projects")
        
        .field("technical_skills")
            .desc("What programming languages and technologies are you proficient in?")
            .hint("Please mention specific languages, frameworks, and tools")
        
        # Confidential field - tracked but never mentioned
        .field("has_mentored")
            .desc("Gives specific evidence of professionally mentoring junior colleagues")
            .confidential()
            .as_bool()
        
        # Another confidential field
        .field("shows_leadership")
            .desc("Demonstrates leadership qualities or initiatives")
            .confidential()
            .as_bool()
        
        # Conclusion field - assessed at the end
        .field("preparedness")
            .desc("Evaluate candidate's preparation: research, questions, examples")
            .conclude()
            .as_one.assessment("unprepared", "somewhat prepared", "well prepared", "exceptionally prepared")
        
        .field("cultural_fit")
            .desc("Assessment of cultural fit based on values and communication style")
            .conclude()
            .as_one.assessment("poor fit", "potential fit", "good fit", "excellent fit")
        
        .build())


def run_interactive(interview):
    """Run the interview interactively."""
    thread_id = f"job-interview-{os.getpid()}"
    print(f'Starting job interview (thread: {thread_id})')
    print("=" * 60)
    print("Software Engineer Position Interview")
    print("=" * 60)
    
    interviewer = Interviewer(interview, thread_id=thread_id)
    
    user_input = None
    print("\n[Interview begins]\n")
    
    while not interview._done:
        message = interviewer.go(user_input)
        
        if message:
            print(f"\nInterviewer: {message}")
        
        if not interview._done:
            try:
                user_input = input("\nCandidate: ").strip()
            except (KeyboardInterrupt, EOFError):
                print("\n[Interview ended by user]")
                break
    
    if interview._done:
        print("\n[Interview complete - Thank you for your time]")
    
    return interview._done


def run_automated(interview):
    """Run with prefab inputs for demonstration."""
    prefab_inputs = [
        "I've spent 5 years in finance, but I've been passionate about programming since college. "
        "I built an automated trading system in Python that saved our team 20 hours per week. "
        "I also created a risk analysis dashboard using React and D3.js.",
        
        "I'm proficient in Python, JavaScript, and SQL. I've worked extensively with React, Node.js, "
        "and Django. I also have experience with Docker, AWS, and CI/CD pipelines using GitHub Actions.",
        
        "In my previous role, I regularly mentored junior analysts on Python programming and data analysis. "
        "I also led a cross-functional team to implement our new reporting system.",
        
        "I've researched your company's focus on fintech innovation and your recent Series B funding. "
        "I'm particularly excited about your API-first approach and how my background in both finance "
        "and engineering could contribute to bridging technical and business requirements. "
        "What are the biggest technical challenges your team is currently facing?"
    ]
    
    thread_id = f"job-demo-{os.getpid()}"
    print(f'Running automated demo (thread: {thread_id})')
    print("=" * 60)
    print("Software Engineer Position Interview (Demo)")
    print("=" * 60)
    
    interviewer = Interviewer(interview, thread_id=thread_id)
    
    input_iter = iter(prefab_inputs)
    user_input = None
    
    print("\n[Interview begins]\n")
    
    while not interview._done:
        message = interviewer.go(user_input)
        
        if message:
            print(f"\nInterviewer: {message}")
        
        if not interview._done:
            try:
                user_input = next(input_iter)
                print(f"\nCandidate: {user_input}")
            except StopIteration:
                print("\n[Demo inputs exhausted]")
                break
    
    if interview._done:
        print("\n[Interview complete - Thank you for your time]")
    
    return interview._done


def display_results(interview):
    """Display the interview assessment."""
    print("\n" + "=" * 60)
    print("INTERVIEW ASSESSMENT SUMMARY")
    print("-" * 60)
    
    # Regular fields
    if interview.experience:
        print("✓ Experience shared: Yes")
    
    if interview.technical_skills:
        print("✓ Technical skills discussed: Yes")
    
    # Confidential fields (for internal assessment)
    print("\n--- Confidential Assessment ---")
    
    if hasattr(interview, 'has_mentored') and interview.has_mentored is not None:
        mentoring = "Yes" if interview.has_mentored.as_bool else "No"
        print(f"Has mentored others: {mentoring}")
    
    if hasattr(interview, 'shows_leadership') and interview.shows_leadership is not None:
        leadership = "Yes" if interview.shows_leadership.as_bool else "No"
        print(f"Shows leadership qualities: {leadership}")
    
    # Conclusion fields
    print("\n--- Final Evaluation ---")
    
    if interview.preparedness:
        print(f"Preparation level: {interview.preparedness}")
    
    if interview.cultural_fit:
        print(f"Cultural fit: {interview.cultural_fit}")
    
    # Check activated traits
    print("\n--- Candidate Profile ---")
    
    traits = interview._chatfield['roles']['bob'].get('possible_traits', {})
    if traits.get('career-changer', {}).get('active'):
        print("• Career changer: Yes")
    
    if traits.get('senior-level', {}).get('active'):
        print("• Senior level: Yes")
    
    print("=" * 60)


def main():
    """Main entry point."""
    parser = argparse.ArgumentParser(description='Job Interview Example')
    parser.add_argument('--auto', action='store_true',
                        help='Run with automated demo inputs')
    args = parser.parse_args()
    
    # Load environment variables
    dotenv.load_dotenv(override=True)
    
    # Check for API key
    if not os.getenv('OPENAI_API_KEY'):
        print("Error: OPENAI_API_KEY not found in environment")
        print("Please set your OpenAI API key in .env file")
        sys.exit(1)
    
    # Create and run the interview
    interview = create_job_interview()
    
    if args.auto:
        completed = run_automated(interview)
    else:
        completed = run_interactive(interview)
    
    # Display results if completed
    if completed:
        display_results(interview)
    else:
        print("\n[Interview not completed]")


if __name__ == "__main__":
    main()